### ISSUE TITLE
aiohttp server hangs when nginx try to mixin plain http request with failed websocket connection upgrade in one tcp stream.

### ISSUE THREAD

**Issue by zTrix** [aio-libs/aiohttp#3380]
## Long story short

aiohttp server hangs when nginx try to mixin plain http request with failed websocket connection upgrade in one tcp stream.

The following tcpdump capture shows what happens in detail.

![deepinscreenshot_select-area_20181105120935](https://user-images.githubusercontent.com/471934/47977582-df21a900-e0f3-11e8-9153-41ecec704508.png)

## Expected behaviour

Server returns something.

## Actual behaviour

Server hangs forever without closing tcp connection. Nginx return 504 Gateway timeout to end user.

## Steps to reproduce

1. write a simple api server using aiohttp (version 3.4.4 in my case), implement some /api/xxx for testing, leave /ws unimplemented.
2. using nginx to reverse proxy to this api server.
3. user open browser targeting a web page which contains both ws request and api request. Then api query will gateway timeout.

![deepinscreenshot_select-area_20181105122712](https://user-images.githubusercontent.com/471934/47978021-493b4d80-e0f6-11e8-9a8c-8cc234193467.png)

Example nginx.conf:

```
    ...
    location /api/ {
        proxy_pass http://server_backend;
        proxy_http_version 1.1;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    }

    location /ws {
        proxy_pass http://server_backend;
        proxy_http_version 1.1;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection "upgrade";
    }
```

aiohttp server serves `/api/` request, but leave `/ws` unimplemented.

Then this issue could be reproduced.

## Your environment

Docker: nginx:latest
Docker: aiohttp3.4.4 on python:3.7-alpine3.7

**Comment by zTrix** [comment]
The weird thing here is why nginx still sends http/1.0 request to upstream backend server while `proxy_http_version 1.1;` specified in nginx.conf.

**Comment by kxepal** [comment]
Your nginx configuration works fine for me for a bit modified `web_ws.py` example. Could you share the code which reproduces the issue? And full nginx config could be useful as well.

**Comment by zTrix** [comment]
[ws-api-mixed-in-one-tcp-stream-hang-bug-demo.tar.gz](https://github.com/aio-libs/aiohttp/files/2547315/ws-api-mixed-in-one-tcp-stream-hang-bug-demo.tar.gz)

Here attached a minimal environment to reproduce this issue.

to start the servers, run
```
$ docker-compose up -d
```

to reproduce the issue, follow the steps in http://your_server/index.html

**Comment by kxepal** [comment]
Thanks! I don't have docker, but without it everything works fine, nothing hangs.

**Comment by zTrix** [comment]
Maybe different nginx version behave differently?

BTW, you need to follow the steps described in index.html:

1. visit http://your_ip/api/null first to see everything ok
2. visit http://your_ip/ws to see a 404 page
3. visit http://your_ip/api/null again, which will get you a 504 Gateway timeout after around 1 minute

It works out on condition that, nginx did try to keepalive with upstream server, send the third request using the same tcp connection used by step 2.

**Comment by zTrix** [comment]
Running the client.py against a golang written demo server works just fine. Here is the demo code:

server.go

```golang
package main

import (
    "fmt"
    "log"
    "net/http"
)

func handler(w http.ResponseWriter, r *http.Request) {
    w.Header().Set("Content-Type", "application/json")
    fmt.Fprintf(w, "{\"err\":null}")
}

func main() {
    http.HandleFunc("/api/null", handler)
    log.Fatal(http.ListenAndServe(":9211", nil))
}
```

The tcpdump screen capture below:

![deepinscreenshot_select-area_20181105174101](https://user-images.githubusercontent.com/471934/47990239-00998980-e122-11e8-9682-7de88e0ad05c.png)

**Comment by kxepal** [comment]
> Maybe different nginx version behave differently?

1.14.0 on my side.

> BTW, you need to follow the steps described in index.html:

Yes, I followed those instructions and repeat them several times.

**Comment by kxepal** [comment]
> demo.tar.gz

This perfectly reproduces the issue. Thanks!

**Comment by kxepal** [comment]
Meanwhile issue couldn't be reproduced with uvloop. Hm.

**Comment by asvetlov** [comment]
Not yet

**Comment by gjcarneiro** [comment]
I've had this bug in production, with latest aiohttp.  Disabling upstream http connection keepalive in nginx fixed it.

> Meanwhile issue couldn't be reproduced with uvloop. Hm.

Not so.  My production code runs aiohttp with uvloop, and still experienced this issue.

**Comment by asvetlov** [comment]
I think the bug is fixed by #3971
